lib/pmdtester/builders/html_report_builder.rb

```diff
@@ -122,55 +118,60 @@ def build_violations_section(doc, violation_diffs)
 
     def build_violation_table(doc, key, value)
       doc.table(class: 'bodyTable', border: '0') do
-        doc.thead do
-          doc.tr do
-            doc.th
-            doc.th 'priority'
-            doc.th 'Rule'
-            doc.th 'Message'
-            doc.th 'Line'
-          end
+        build_violation_table_head(doc)
+        build_violation_table_body(doc, key, value)
+      end
+    end
+
+    def build_violation_table_head(doc)
+      doc.thead do
+        doc.tr do
+          doc.th
+          doc.th 'priority'
+          doc.th 'Rule'
+          doc.th 'Message'
+          doc.th 'Line'
         end
+      end
+    end
 
-        doc.tbody do
-          a_index = 1
-          value.each do |pmd_violation|
-            doc.tr(class: pmd_violation.branch == 'base' ? 'a' : 'b') do
-              # The anchor
-              doc.td do
-                doc.a(id: "A#{a_index}", href: "#A#{a_index}") do
-                  doc.text '#'
-                end
-              end
-
-              a_index += 1
-              violation = pmd_violation.violation
-
-              # The priority of the rule
-              doc.td violation['priority']
-
-              # The rule that trigger the violation
-              doc.td do
-                doc.a(href: (violation['externalInfoUrl']).to_s) do
-                  doc.text violation['rule']
-                end
-              end
-
-              # The violation message
-              doc.td violation.text
-
-              # The begin line of the violation
-              line = violation['beginline']
-
-              # The link to the source file
-              doc.td do
-                link = get_link_to_source(violation, key)
-                doc.a(href: link.to_s) do
-                  doc.text line
-                end
-              end
-            end
-          end
+    def build_violation_table_body(doc, key, value)
+      doc.tbody do
+        a_index = 1
+        value.each do |pmd_violation|
+          build_violation_table_raw(a_index, key, pmd_violation, doc)
+          a_index += 1
+        end
+      end
+    end
+
+    def build_violation_table_raw(a_index, key, pmd_violation, doc)
```

Comment:
The name should be actually "row" (not "raw").

I'd put the doc parameter as first, since for the other methods, the document is always the first parameter.
"key" will be the second parameter then (e.g. similar to the method "build_violation_table_body", or "build_violation_table").